app/ts/components/HashRouter.tsx

```diff
@@ -0,0 +1,81 @@
+import { signal } from '@preact/signals'
+import { FunctionalComponent, VNode } from 'preact'
+import { useEffect } from 'preact/hooks'
+
+export type RouteProps = {
+	path: string
+	component: FunctionalComponent
+}
+
+export const Route = ({ component: Component }: RouteProps) => <Component />
+
+export type RouterProps = {
+	children: object[]
+	onHashChange?: (event: Event) => void
+}
+
+const activeRoute = signal<VNode<RouteProps> | undefined>(undefined)
+
+export const Router = ({ children, onHashChange }: RouterProps) => {
+	function getActiveRoute() {
+		return children.filter(isRouteComponent).find(child => matchPattern(child.props.path, window.location.hash))
+	}
+
+	activeRoute.value = getActiveRoute()
+
+	useHashChangeEffect((event: Event) => {
+		activeRoute.value = getActiveRoute()
+		onHashChange?.(event)
+	})
+
+	return <main>{activeRoute.value}</main>
+}
+
+export function useParams() {
+	if (!activeRoute.value) return
+	return matchPath(activeRoute.value.props.path, window.location.hash)
+}
+
+function useHashChangeEffect(handler: (event: Event) => void) {
+	useEffect(() => {
+		window.addEventListener('hashchange', handler)
+		return () => window.removeEventListener('hashchange', handler)
+	}, [])
+}
+
+function isRouteComponent(child: object): child is VNode<Parameters<typeof Route>[0]> {
```

Comment:
Thinking more on this, I think we should only assert that we have a `VNode` with a `path` property.  Are there some checks we can do to verify that we actually have a VNode?  Right now really all we are doing is asserting that we are returning something with a `path: string` property.
```suggestion
function isRouteComponent(child: object): child is VNode<{ path: string }> {
```

Reply:
or would you suggest we just add the `component` check as well so it's exclusively for the `Route` component? the function name already suggests it should be

Reply:
> Are there some checks we can do to verify that we actually have a VNode
The only required prop for VNode is `children: ComponentChildren` which could also be another `VNode<any>`

Reply:
I guess just add a check for `children` then.  😖